File: go/general/589. N叉树的前序遍历.go
package main

var res []int

func preorder(root *Node) []int {
	res = []int{}
	dfs(root)
	return res
}

func dfs(root *Node) {
	if root != nil {
		res = append(res, root.Val)
		for _, v := range root.Children {
			dfs(v)
		}
	}
}

File: python/589. N叉树的前序遍历.py
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children

from typing import List

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        res = []
        def dfs(root):
            if root:
                res.append(root.val)
                for node in root.children:
                    dfs(node)
        dfs(root)
        return res

# 迭代
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        stack = [root]
        res = []
        while stack:
            node = stack.pop()
            if node:
                res.append(node.val)
                for node in node.children[::-1]:
                    stack.append(node)
        return res